lib/fs.js

```diff
@@ -395,12 +395,32 @@ function readFileAfterClose(err) {
   else
     buffer = context.buffer;
 
-  if (context.encoding)
-    buffer = buffer.toString(context.encoding);
+  if (err) return callback(err, buffer);
 
-  callback(err, buffer);
+  if (context.encoding) {
+    const result = tryToString(buffer, context.encoding);
+    if (result.error)
+      return callback(result.error, buffer);
```

Comment:
However, since you're returning in either case, why not simply pass the callback to the tryToString method...? 

```js
if (context.encoding)
  return tryToString(buffer, context.encoding, callback);
//...

function tryToString(buf, encoding, callback) {
  try {
    return callback(null, buf.toString(encoding));
  }
  catch (err) {
    return callback(err, buf);
  }
}
```